src/server/pkg/ppsutil/util.go

```diff
@@ -283,31 +304,57 @@ func NewPipelineManifestReader(path string) (result *PipelineManifestReader, ret
 				retErr = err
 			}
 		}()
-		rawBytes, err := ioutil.ReadAll(resp.Body)
-		if err != nil {
-			return nil, err
-		}
-		pipelineReader = io.TeeReader(strings.NewReader(string(rawBytes)), &result.buf)
+		pipelineBytes, err = ioutil.ReadAll(resp.Body)
 	} else {
-		rawBytes, err := ioutil.ReadFile(path)
-		if err != nil {
-			return nil, err
-		}
-
-		pipelineReader = io.TeeReader(strings.NewReader(string(rawBytes)), &result.buf)
+		pipelineBytes, err = ioutil.ReadFile(path)
+	}
+	if err != nil {
+		return nil, err
+	}
+	result = &PipelineManifestReader{
+		ds: []pipelineDecoder{
+			// create two independent readers for the two decoders so that one decoder
+			// reading doesn't affect the other one.
+			yaml.NewDecoder(bytes.NewReader(pipelineBytes)),
+			json.NewDecoder(bytes.NewReader(pipelineBytes)),
+		},
 	}
-	result.decoder = json.NewDecoder(pipelineReader)
 	return result, nil
 }
 
 // NextCreatePipelineRequest gets the next request from the manifest reader.
 func (r *PipelineManifestReader) NextCreatePipelineRequest() (*ppsclient.CreatePipelineRequest, error) {
 	var result ppsclient.CreatePipelineRequest
-	if err := jsonpb.UnmarshalNext(r.decoder, &result); err != nil {
-		if err == io.EOF {
-			return nil, err
+	retErr := NoParseErr // set to nil on 1st successful parse
+	dest := interface{}(&result)
+	// Call Decode() on every d.ds. After the first success, continue calling
+	// decode on other decoders (to keep them in sync) but throw other results
+	// away.
+	for i := 0; i < len(r.ds); {
```

Comment:
this loop seems overly complicated for what it's trying to do... especially when we only ever expect two decoders

Reply:
I tried to shrink the code a bit by inlining comments and such, but I'm not quite sure what an overall shorter way to do this might be. Maybe the error semantics are too complicated: my goal was to return the error from the last parser to fail (so JSON specs get JSON errors and YAML specs get YAML errors, instead of something silly like "your YAML spec is not JSON" when the real problem is that the YAML spec has a tiny error".

When I just store e.g. `jsonDecoder` and `yamlDecoder`, it ended up being the same number of lines (after inlining comments) and IMO more confusing, but if you prefer it I'm happy to go with it:
```
func (r *PipelineManifestReader) NxtCreatePipelineRequest() (*ppsclient.CreatePipelineRequest, error) {
	if d.jsonErr != nil && d.yamlErr != nil {
		return errors.New("no decoders are able to parse pipeline spec")
	}
	var result ppsclient.CreatePipelineRequest
	dest := interface{}(&result)
	if d.jsonErr != nil {
		d.jsonErr = d.jsonDecoder.Decode(dest)
		if d.yamlErr == nil && d.jsonErr == nil {
			dest = &map[string]interface{}{} // deposit YAML parse into garbage map
		} else if d.yamlErr != nil && d.jsonErr != nil {
			// yaml parse failed earlier--json err is most recent
			if d.jsonErr == io.EOF {
				return io.EOF
			}
			return nil, fmt.Errorf("malformed pipeline spec: %s", d.jsonErr)
		}
	}
	if d.yamlErr != nil {
		d.yamlErr = d.yamlDecoder.Decode(dest)
		if d.jsonErr != nil && d.yamlErr != nil {
			if d.yamlErr == io.EOF {
				return io.EOF
			}
			return nil, fmt.Errorf("malformed pipeline spec: %s", d.yamlErr)
		}
	}

	return nil, fmt.Errorf("malformed pipeline spec: %s", retErr)
	return &result, nil
}
```

Reply:
Another idea I had which is pretty hacky but might simplify the code is to look for the first non-space character, and if it's `{` use the JSON parser, otherwise use the YAML parser?

Reply:
I think the new version is alright